### PR TITLE
Support PEP560 for extension class

### PR DESCRIPTION
**Pull request by KmolYuan**
#3764 Support `__class_getitem__` and add an error for `__init_subclass__`.

### CODE REVIEW

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -2965,10 +2966,12 @@ def analyse_declarations(self, env):
...
             self.is_staticmethod = False
 
-        if env.is_py_class_scope:
-            if self.name == '__new__':
+        if env.is_py_class_scope or env.is_c_class_scope:
+            if self.name == '__new__' and env.is_py_class_scope:
                 self.is_staticmethod = True
-            elif not self.is_classmethod and self.name in IMPLICIT_CLASSMETHODS:
+            elif self.name == INHERITANCE_HOOK and env.is_c_class_scope:
+                error(self.pos, "inheritance argument is not supported by extension class")
```
Comment (da-woods) [inline Cython/Compiler/Nodes.py:2973]: I'd make this more explicit: `"__init_subclass__ is not supported by extensions class"`

File: Cython/Compiler/Nodes.py [hunk 2]
```diff
@@ -39,7 +39,8 @@
     _py_int_types = (int, long)
 
 
-IMPLICIT_CLASSMETHODS = {"__init_subclass__", "__class_getitem__"}
+INHERITANCE_HOOK = '__init_subclass__'
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:42]: Please inline this constant. It's not clear from the name what it does. Seeing `"__init_subclass__"` allows me to search for it, seeing `INHERITANCE_HOOK` requires me to look up its value before I can search for it.

**Comment by scoder** [comment]
> The ones from #3518 don't look great to copy

It's a bit annoying that they use locally defined nested classes, but apart from that, we're touching a different class implementation here, so it needs as thorough testing as Python classes.

**Comment by KmolYuan** [comment]
I guess some statically defined extension classes should be tested here?
(excluding mixed with `__init_subclass__` and meta class)

And I don't know how to test compiler error, are there ready-made examples for reference?

**Comment by scoder** [comment]
> I guess some statically defined extension classes should be tested here?

I'd say you can just copy `test_genericclass.py` to a new file (say, `test_genericclass_exttype.pyx`) and turn all Python test classes into extension types there. (Shouldn't we add a `@cython.auto_cclass` directive at some point?)

> And I don't know how to test compiler error, are there ready-made examples for reference?

Sure, lots of them in `tests/errors/`. See the [docs](https://github.com/cython/cython/wiki/HackerGuide#the-test-suite).

File: tests/run/test_genericclass_exttype.pyx [hunk 3]
```diff
@@ -0,0 +1,99 @@
+# mode: run
+# tag: pure3.7
```
Comment (scoder) [inline tests/run/test_genericclass_exttype.pyx:2]: This isn't needed since it's not a pure Python (`.py`) test but a `.pyx` test, so Python won't see it anyway.

File: Cython/Compiler/Nodes.py [hunk 4]
```diff
@@ -39,7 +39,7 @@
     _py_int_types = (int, long)
 
 
-IMPLICIT_CLASSMETHODS = {"__init_subclass__", "__class_getitem__"}
+IMPLICIT_CLASSMETHODS = {'__init_subclass__', '__class_getitem__'}
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:42]: Looks like an unnecessary change.

**Comment by scoder** [comment]
Thanks!
